text/0000-duration-reform.md

```diff
@@ -0,0 +1,162 @@
+> parameter as a [`DWORD`][msdn-dword], which is a `u32`. Some Windows
+> APIs use "ticks" or 100-nanosecond units.
+
+[iocp-ms-example]: https://msdn.microsoft.com/en-us/library/windows/desktop/aa364986%28v=vs.85%29.aspx
+[msdn-dword]: https://msdn.microsoft.com/en-us/library/cc230318.aspx
+
+In light of that, this proposal has two primary goals:
+
+* to define a type that can describe portable timeouts for cross-
+  platform APIs
+* to describe what should happen if a large `Duration` is passed into
+  an API that does not accept timeouts that large
+
+In general, this proposal considers it acceptable to reduce the granularity of timeouts (eliminating nanosecond granularity if only milliseconds are supported) and to truncate very large timeouts.
+
+This proposal retains the two fields in the existing `Duration`:
+
+* a `u64` of seconds
+* a `u32` of additional nanosecond precision
+
+Timeout APIs defined in terms of milliseconds will truncate `Duration`s that are more than `u32::MAX` in milliseconds, and will reduce the granularity of the nanosecond field.
+
+> A `u32` of milliseconds supports a timeout longer than 45 days.
+
+Future APIs to support a broader set of [Durations][joda-duration] APIs, a [Period][joda-period] and [Instant][joda-instant] type, as well as coercions between these types, would be useful, compatible follow-ups to this RFC.
+
+[joda-duration]: http://www.joda.org/joda-time/key_duration.html
+[joda-period]: http://www.joda.org/joda-time/key_period.html
+[joda-instant]: http://www.joda.org/joda-time/key_instant.html
+
+# Detailed design
+
+A `Duration` represents a period of time represented in terms of nanosecond granularity. It has `u64` seconds and an additional `u32` nanoseconds. There is no concept of a negative `Duration`.
+
+> A negative `Duration` has no meaning for many APIs that may wish
+> to take a `Duration`, which means that all such APIs would need
+> to decide what to do when confronted with a negative `Duration`.
+> As a result, this proposal focuses on the predominant use-cases for
+> `Duration`, where unsigned types remove a number of caveats and
+> ambiguities.
+
+```rust
+pub struct Duration {
+  secs: u64,
+  nanos: u32 // may not be more than 1 billion
+}
+
+impl Duration {
+    /// create a Duration from a number of seconds and an
+    /// additional nanosecond precision
+    pub fn new(secs: u64, nanos: u32) -> Timeout;
+
+    /// create a Duration from a number of seconds
+    pub fn from_secs(secs: u64) -> Timeout;
+
+    /// create a Duration from a number of milliseconds
+    pub fn from_millis(millis: u64) -> Timeout;
+
+    /// the number of seconds represented by the Timeout
+    pub fn secs(self) -> u64;
+
+    /// the number of additional nanosecond precision
+    pub fn nanos(self) -> u32;
+}
+```
+
+When `Duration` is used with a system API that expects `u32` milliseconds, the nanosecond precision is dropped, and the time is truncated to `u32::MAX`.
```

Comment:
I think it's just poorly worded - it means drop the nanosecond precision of the combined duration, leaving only millisecond precision, not drop the nanosecond field and only use the second field.

What's the reasoning behind rounding down rather than up? It's just that APIs such as "Sleep" generally have an "at least" guarantee rather than an "at most" guarantee, ie. if you passed in 5ms, it would sleep at least 5ms and possible longer, but if you pass in 4.9ms using this API, it may actually sleep for a shorter time than was specified, which might be confusing.

I'm not sure what the best answer is, since clearly some APIs have the opposite guarantee, but it's something that should be mentioned IMO.